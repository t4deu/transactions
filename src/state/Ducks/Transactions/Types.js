// @flow
import type { Store as ReduxStore, Dispatch as ReduxDispatch } from 'redux';

export const ADD_TRANSACTION = 'TRANSACTION/ADD_TRANSACTION';
export const SET_SORT_ORDER = 'TRANSACTION/SET_SORT_ORDER';

export type TransactionType = 'credit' | 'debit';
export type DateString = string;
export type SortOrder = 'DESC' | 'ASC';

export type Transaction = {
  +id: number,
  +type: TransactionType,
  +amount: number,
  subtotal: number,
  +description: string,
  +createdAt: DateString,
};

export type TransactionPayload = {
  +type: TransactionType,
  +amount: string,
  +description: string,
};

export type Status = {
  +total: number,
  +creditTotal: number,
  +debitTotal: number,
  +isSaver: boolean,
};

export type Transactions = Array<Transaction>;

export type transactionsState = {
  +transactions: Transactions,
  +sortOrder: SortOrder,
};

export type Action =
  | {
      type: 'TRANSACTION/ADD_TRANSACTION',
      payload: TransactionPayload,
    }
  | { type: 'TRANSACTION/SET_SORT_ORDER', +sort: SortOrder };

export type Store = ReduxStore<transactionsState, Action>;
export type Dispatch = ReduxDispatch<Action>;
