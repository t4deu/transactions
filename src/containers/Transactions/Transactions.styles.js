export default {
  container: {
    flex: 1,
  },
  bar: {
    height: 24,
    backgroundColor: '#1e88e5',
  },
};
