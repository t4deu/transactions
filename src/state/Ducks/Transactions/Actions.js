// @flow
import { ADD_TRANSACTION, SET_SORT_ORDER } from './Types';
import type { Action, SortOrder, TransactionType } from './Types';

export const addTransaction = (
  type: TransactionType,
  amount: number,
  description: string,
): Action => ({
  type: ADD_TRANSACTION,
  payload: {
    type,
    amount,
    description,
  },
});

export const setSortOrder = (sortOrder: SortOrder): Action => ({
  type: SET_SORT_ORDER,
  sort: sortOrder,
});
