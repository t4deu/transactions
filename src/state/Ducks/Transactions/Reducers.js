// @flow
import { combineReducers } from 'redux';

import { nextId, currentTime, parseNumber } from './Utils';

import { ADD_TRANSACTION, SET_SORT_ORDER } from './Types';
import type { Transactions, Transaction, Action, SortOrder } from './Types';

/* state shape
transactionsState {
  transactions: [],
  sortOrder: "DESC"
} */

function createTransaction(state: Transactions, action): Transaction {
  const { type, amount, description } = action.payload;
  const validAmount = parseNumber(amount);
  const signedAmount = type === 'debit' ? -validAmount : validAmount;

  return {
    id: nextId(state),
    createdAt: currentTime(),
    amount: signedAmount,
    type,
    description,
  };
}

const transactions = (state: Transactions = [], action: Action): Transactions => {
  switch (action.type) {
    case ADD_TRANSACTION:
      return [...state, createTransaction(state, action)];
    default:
      return state;
  }
};

const sortOrder = (state: SortOrder = 'DESC', action: Action): SortOrder => {
  switch (action.type) {
    case SET_SORT_ORDER:
      return action.sort;
    default:
      return state;
  }
};

const transactionsReducer = combineReducers({
  transactions,
  sortOrder,
});

export default transactionsReducer;
