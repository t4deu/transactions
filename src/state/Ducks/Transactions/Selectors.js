// @flow
import currency from 'currency.js';

import type { Transactions, SortOrder, Status } from './Types';

export const getSortedTransactions = (
  transactions: Transactions,
  order: SortOrder,
): Transactions => {
  switch (order) {
    case 'ASC':
      return transactions.sort((a, b) => a.id - b.id);
    case 'DESC':
      return transactions.sort((a, b) => b.id - a.id);
    default:
      throw new Error(`Unknown sort order: ${order}`);
  }
};

export const getTransactionsWithSubTotal = (transactions: Transactions): Transactions => {
  return transactions.map((transaction) => {
    const subtotal = transactions
      .filter(item => transaction.id >= item.id)
      .map(item => item.amount)
      .reduce((total, current) => currency(total).value + currency(current).value, currency(0));

    return {
      ...transaction,
      subtotal,
    };
  }).sort((a, b) => a.id - b.id);
};

export const getTransactionsStatus = (transactions: Transactions): Status => {
  const credits = transactions.filter(transaction => transaction.type === 'credit');
  const debits = transactions.filter(transaction => transaction.type === 'debit');
  const creditTotal = credits.reduce(
    (total, transaction) =>
      // avoid float precision issues
      total.add(transaction.amount),
    currency(0),
  );
  const debitTotal = debits.reduce(
    (total, transaction) =>
      // avoid float precision issues
      total.add(transaction.amount),
    currency(0),
  );
  const total = currency(creditTotal).add(debitTotal);
  const isSaver = total.value > 0;

  return {
    creditTotal: creditTotal.value,
    debitTotal: debitTotal.value,
    total: total.value,
    isSaver,
  };
};
