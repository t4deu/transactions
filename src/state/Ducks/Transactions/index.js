import transactionsReducer from './Reducers';

export { default as transactionsSelectors } from './Selectors';
export { default as transactionsTypes } from './Types';

export default transactionsReducer;
