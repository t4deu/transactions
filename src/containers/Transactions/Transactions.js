// @flow
// external imports
import React, { Component } from 'react';
import { View, StatusBar } from 'react-native';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Toolbar } from 'react-native-material-ui';
// types
import type {
  SortOrder,
  TransactionType,
  TransactionPayload,
  Status as CurrentStatus,
  Transactions as TransactionsList,
} from '../../state/Ducks/Transactions/Types';
// actions & selectors
import { addTransaction, setSortOrder } from '../../state/Ducks/Transactions/Actions';
import {
  getSortedTransactions,
  getTransactionsStatus,
} from '../../state/Ducks/Transactions/Selectors';
// containers & components
import Controls from '../../components/Controls/Controls';
import FirstTransaction from '../../components/FirstTransaction/FirstTransaction';
import TransactionForm from '../../components/Form/Form';
import Status from '../../components/Status/Status';
import Timeline from '../../components/Timeline/Timeline';
// styles
import styles from './Transactions.styles';

type Props = {
  transactions: TransactionsList,
  currentStatus: CurrentStatus,
  hasTransactions: boolean,
  addTransaction: Function,
  setSortOrder: Function,
  sortOrder: SortOrder,
};

type State = {
  formActive: boolean,
  formType: TransactionType,
};

export class Transactions extends Component<Props, State> {
  state = {
    formActive: false,
    formType: 'credit',
  };

  addTransaction = (payload: TransactionPayload) => {
    this.props.addTransaction(payload);
    this.closeForm();
  };

  addCredit = () => {
    this.setState({
      formActive: true,
      formType: 'credit',
    });
  };

  addDebit = () => {
    this.setState({
      formActive: true,
      formType: 'debit',
    });
  };

  closeForm = () => {
    this.setState({
      formActive: false,
    });
  };

  toogleSortOrder = () => {
    if (this.props.sortOrder === 'DESC') {
      this.props.setSortOrder('ASC');
    } else {
      this.props.setSortOrder('DESC');
    }
  }

  tollbarAction() {
    if (!this.props.hasTransactions) {
      return {};
    }

    return this.props.sortOrder === 'ASC'
      ? {
        actions: ['file-upload'],
      }
      : {
        actions: ['file-download'],
      };
  }

  renderStatus() {
    return <Status currentStatus={this.props.currentStatus} />;
  }

  renderContent() {
    if (!this.props.hasTransactions) {
      return <FirstTransaction onAddCredit={this.addCredit} onAddDebit={this.addDebit} />;
    }

    return (
      <View style={styles.container}>
        <Timeline header={this.renderStatus()} transactions={this.props.transactions} />
        <Controls onAddCredit={this.addCredit} onAddDebit={this.addDebit} />
      </View>
    );
  }

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.bar} />
        <Toolbar
          centerElement="Transações"
          rightElement={this.tollbarAction()}
          onRightElementPress={this.toogleSortOrder}
        />
        {this.renderContent()}
        <TransactionForm
          type={this.state.formType}
          active={this.state.formActive}
          onBack={this.closeForm}
          onSubmit={this.addTransaction}
        />
      </View>
    );
  }
}

const stateToProps = (state) => {
  const { transactions, sortOrder } = state.transactionsState;
  return {
    transactions: getSortedTransactions(transactions, sortOrder),
    hasTransactions: transactions.length > 0,
    currentStatus: getTransactionsStatus(transactions),
    sortOrder,
  };
};

const dispatchToProps = dispatch =>
  bindActionCreators(
    {
      addTransaction,
      setSortOrder,
    },
    dispatch,
  );

const SortedTransactions = connect(stateToProps, dispatchToProps)(Transactions);

export default SortedTransactions;
