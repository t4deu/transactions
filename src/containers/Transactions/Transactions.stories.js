import React from 'react';
import { storiesOf } from '@storybook/react-native';

import { Transactions } from './Transactions';

storiesOf('Transactions', module)
  .add('Saver with transactions', () => (
    <Transactions
      currentStatus={{
        debitTotal: -200,
        creditTotal: 4000,
        total: 3800,
        isSaver: true,
      }}
      hasTransactions
      sortOrder="ASC"
      transactions={[
        { id: 1, type: 'credit', amount: 10.50, description: 'Depósito poupança', createdAt: '2018-03-20' },
        { id: 2, type: 'debit', amount: -60, description: 'Confraria da Pizza', createdAt: '2018-03-20' },
        { id: 3, type: 'credit', amount: 115.90, description: 'Depósito poupança', createdAt: '2018-03-20' },
        { id: 4, type: 'debit', amount: -13.1, description: 'Confraria da Pizza', createdAt: '2018-03-20' },
        { id: 5, type: 'debit', amount: -140, description: 'Depósito poupança', createdAt: '2018-03-20' },
        { id: 6, type: 'credit', amount: 10, description: 'Confraria da Pizza', createdAt: '2018-03-20' },
      ]}
    />
  ))
  .add('Saver with single transaction', () => (
    <Transactions
      currentStatus={{
        debitTotal: -200,
        creditTotal: 4000,
        total: 3800,
        isSaver: true,
      }}
      hasTransactions
      sortOrder="ASC"
      transactions={[
        { id: 1, type: 'credit', amount: 10.50, description: 'Depósito poupança', createdAt: '2018-03-20' },
      ]}
    />
  ))
  .add('Spender with transactions', () => (
    <Transactions
      currentStatus={{
        debitTotal: 200,
        creditTotal: -4000,
        total: -3800,
        isSaver: false,
      }}
      hasTransactions
      sortOrder="ASC"
      transactions={[
        { id: 1, type: 'credit', amount: 10.50, description: 'Depósito poupança', createdAt: '2018-03-20' },
        { id: 2, type: 'debit', amount: -60, description: 'Confraria da Pizza', createdAt: '2018-03-20' },
        { id: 3, type: 'credit', amount: 115.90, description: 'Depósito poupança', createdAt: '2018-03-20' },
        { id: 4, type: 'debit', amount: -13.1, description: 'Confraria da Pizza', createdAt: '2018-03-20' },
        { id: 5, type: 'debit', amount: -140, description: 'Depósito poupança', createdAt: '2018-03-20' },
        { id: 6, type: 'credit', amount: 10, description: 'Confraria da Pizza', createdAt: '2018-03-20' },
      ]}
    />
  ))
  .add('Spender reverse order', () => (
    <Transactions
      currentStatus={{
        debitTotal: 200,
        creditTotal: -4000,
        total: -3800,
        isSaver: false,
      }}
      hasTransactions
      sortOrder="DESC"
      transactions={[
        { id: 1, type: 'credit', amount: 10.50, description: 'Depósito poupança', createdAt: '2018-03-20' },
        { id: 2, type: 'debit', amount: -60, description: 'Confraria da Pizza', createdAt: '2018-03-20' },
      ]}
    />
  ))
  .add('Without transactions', () => (
    <Transactions
      currentStatus={{
        debitTotal: 0,
        creditTotal: 0,
        total: 0,
        isSaver: true,
      }}
      transactions={[]}
    />
  ));
