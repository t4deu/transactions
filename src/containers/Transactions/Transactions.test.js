import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Transactions } from './Transactions';

describe('<Transactions>', () => {
  it('renders multiple transactions', () => {
    const component = shallow(<Transactions
      currentStatus={{
          debitTotal: -200,
          creditTotal: 4000,
          total: 3800,
          isSaver: true,
        }}
      hasTransactions
      sortOrder="ASC"
      transactions={[
          {
            id: 1,
            type: 'credit',
            amount: 10.5,
            description: 'Depósito poupança',
            createdAt: '2018-03-20',
          },
          {
            id: 2,
            type: 'debit',
            amount: -60,
            description: 'Confraria da Pizza',
            createdAt: '2018-03-20',
          },
        ]}
    />);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('renders single transaction', () => {
    const component = shallow(<Transactions
      currentStatus={{
          debitTotal: -200,
          creditTotal: 4000,
          total: 3800,
          isSaver: true,
        }}
      hasTransactions
      sortOrder="ASC"
      transactions={[
          {
            id: 1,
            type: 'credit',
            amount: 10.5,
            description: 'Depósito poupança',
            createdAt: '2018-03-20',
          },
        ]}
    />);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('renders without transactions', () => {
    const component = shallow(<Transactions transactions={[]} />);

    expect(toJson(component)).toMatchSnapshot();
  });
});
