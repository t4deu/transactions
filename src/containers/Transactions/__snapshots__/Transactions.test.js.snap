// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Transactions> renders multiple transactions 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1e88e5",
        "height": 24,
      }
    }
  />
  <Toolbar
    centerElement="Transações"
    hidden={false}
    isSearchActive={false}
    leftElement={null}
    onLeftElementPress={null}
    onPress={null}
    onRightElementPress={[Function]}
    rightElement={
      Object {
        "actions": Array [
          "file-upload",
        ],
      }
    }
    searchable={null}
    size={24}
    style={Object {}}
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Timeline
      header={
        <Status
          currentStatus={
            Object {
              "creditTotal": 4000,
              "debitTotal": -200,
              "isSaver": true,
              "total": 3800,
            }
          }
        />
      }
      transactions={
        Array [
          Object {
            "amount": 10.5,
            "createdAt": "2018-03-20",
            "description": "Depósito poupança",
            "id": 1,
            "type": "credit",
          },
          Object {
            "amount": -60,
            "createdAt": "2018-03-20",
            "description": "Confraria da Pizza",
            "id": 2,
            "type": "debit",
          },
        ]
      }
    />
    <Controls
      onAddCredit={[Function]}
      onAddDebit={[Function]}
    />
  </View>
  <C
    active={false}
    onBack={[Function]}
    onSubmit={[Function]}
    type="credit"
  />
</View>
`;

exports[`<Transactions> renders single transaction 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1e88e5",
        "height": 24,
      }
    }
  />
  <Toolbar
    centerElement="Transações"
    hidden={false}
    isSearchActive={false}
    leftElement={null}
    onLeftElementPress={null}
    onPress={null}
    onRightElementPress={[Function]}
    rightElement={
      Object {
        "actions": Array [
          "file-upload",
        ],
      }
    }
    searchable={null}
    size={24}
    style={Object {}}
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Timeline
      header={
        <Status
          currentStatus={
            Object {
              "creditTotal": 4000,
              "debitTotal": -200,
              "isSaver": true,
              "total": 3800,
            }
          }
        />
      }
      transactions={
        Array [
          Object {
            "amount": 10.5,
            "createdAt": "2018-03-20",
            "description": "Depósito poupança",
            "id": 1,
            "type": "credit",
          },
        ]
      }
    />
    <Controls
      onAddCredit={[Function]}
      onAddDebit={[Function]}
    />
  </View>
  <C
    active={false}
    onBack={[Function]}
    onSubmit={[Function]}
    type="credit"
  />
</View>
`;

exports[`<Transactions> renders without transactions 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1e88e5",
        "height": 24,
      }
    }
  />
  <Toolbar
    centerElement="Transações"
    hidden={false}
    isSearchActive={false}
    leftElement={null}
    onLeftElementPress={null}
    onPress={null}
    onRightElementPress={[Function]}
    rightElement={Object {}}
    searchable={null}
    size={24}
    style={Object {}}
  />
  <FirstTransaction
    onAddCredit={[Function]}
    onAddDebit={[Function]}
  />
  <C
    active={false}
    onBack={[Function]}
    onSubmit={[Function]}
    type="credit"
  />
</View>
`;
